app/models/name.rb

```diff
@@ -1688,25 +1693,26 @@ def self.parse_author(str)
   def self.parse_group(str, deprecated = false)
     return unless match = GROUP_PAT.match(str)
 
-    # Parse the string without "group"
-    ungrouped_str = str.sub(/\s#{GROUP_ABBR}\b/, "")
-    result = parse_name(ungrouped_str)
+    result = parse_name(str_without_group(str))
     return nil unless result
 
     # Adjust the parsed name
-    result.text_name += " group"
+    group_chunk = GROUP_CHUNK.match(str)
+    group_type = group_type(group_chunk[:group_abbr])
```

Comment:
I will re-write and ping you when I'm done.
(Those two lines are what I meant, but I had reservations about them and `GROUP_CHUNK`. I'm going to re-write to try to make things more transparent.)